### PR TITLE
Auto-skip: ensure we account for +base

### PR DESCRIPTION
**Pull request by mikejholly**
The auto-skip code did not account for the "base" target. Hence, the change to root-level `ARG`s were not breaking the cache. 

https://github.com/earthly/earthly/issues/3895

### CODE REVIEW

File: tests/autoskip/arg-change.earth [hunk 1]
```diff
@@ -0,0 +1,12 @@
+VERSION --build-auto-skip 0.8
+
+ARG VERSION=3.18.6
+FROM alpine:$VERSION
+
+PROJECT earthly-technologies/cloud
```
Comment (idodod) [inline tests/autoskip/arg-change.earth:6]: nit: 
```suggestion
PROJECT earthly-technologies/core
```

or remove completely?

**Review by idodod — APPROVED** [review]
awesome! thanks for the quick turnaround

File: tests/autoskip/arg-change.earth [hunk 2]
```diff
@@ -0,0 +1,10 @@
+VERSION --build-auto-skip 0.8
+
+ARG VERSION=3.18.6
```
Comment (alexcb) [inline tests/autoskip/arg-change.earth:3]: I worry that someone might be tempted to bump this version (thinking it's good to keep up to date); however it won't be obvious that it needs to stay in sync with the corresponding test in the Earthfile.

I have two ideas:

1. change this to `ARG VERSION=PLACEHOLDER_FOR_DEFAULT_VALUE`, and then in the `test-arg-change` target, do a `sed -i "s/PLACEHOLDER_FOR_DEFAULT_VALUE/3.18.6"`.

or

2. instead of changing the versions, just change the image, e.g.

```
ARG IMG=alpine
FROM $IMG
```

then `sed -i "s/alpine/busybox/g" Earthfile`, and then they will just always use the latest.

File: util/deltautil/Earthfile [hunk 3]
```diff
@@ -1,6 +1,7 @@
 VERSION 0.8
 
-FROM github.com/earthly/earthly+base
+# Note: the tag below is needed for auto-skip.
+FROM github.com/earthly/earthly:tags/v0.8.5+base
```
Comment (alexcb) [inline util/deltautil/Earthfile:4]: could we just go back to using `../..+base` instead?

File: tests/autoskip/arg-change.earth (hunk 2 above)
Comment (mikejholly) [reply tests/autoskip/arg-change.earth:3]: Good point. Does a comment or special Renovate ignore flag sound fair? @idodod do you know how to achieve the latter?

File: util/deltautil/Earthfile (hunk 3 above)
Comment (mikejholly) [reply util/deltautil/Earthfile:4]: I think these full target names were added because `deltautil` & `ast` are separate packages. I'm concerned that using relative paths may break something.

Comment (alexcb) [reply util/deltautil/Earthfile:4]: how about using `ARG EARTHLY_GIT_BRANCH` instead?

e.g. 

```
ARG EARTHLY_GIT_BRANCH
FROM github.com/earthly/earthly:$EARTHLY_GIT_BRANCH+base
```

Comment (alexcb) [reply util/deltautil/Earthfile:4]: or `EARTHLY_GIT_HASH`

Comment (mikejholly) [reply util/deltautil/Earthfile:4]: Good idea. Branches don't work with auto-skip, but hashes will.

File: tests/autoskip/arg-change.earth (hunk 2 above)
Comment (mikejholly) [reply tests/autoskip/arg-change.earth:3]: To elaborate, it would be useful to have a way to pin versions without the extra effort.

File: util/deltautil/Earthfile (hunk 3 above)
Comment (mikejholly) [reply util/deltautil/Earthfile:4]: Actually, the hash will break local runs for unpushed commits. :thinking:

File: tests/autoskip/arg-change.earth (hunk 2 above)
Comment (idodod) [reply tests/autoskip/arg-change.earth:3]: We don't use Renovate (yet) to upgrade the images in Earthfiles but once we do it shouldn't be a problem for it to ignore this specific image version. 

To Alex's point though, Someone (like me:)) might still feel the urge to upgrade it manually. Maybe using a digest might make it clearer that it shouldn't be upgraded?

File: util/deltautil/Earthfile (hunk 3 above)
Comment (idodod) [reply util/deltautil/Earthfile:4]: When running locally wouldn't EARTHLY_GIT_HASH point to the local hash that might not exist in the remote repo yet?

Comment (idodod) [reply util/deltautil/Earthfile:4]: If we use the tag/version, we can configure Renovate to upgrade the version together with all the other places in the repo.

Comment (mikejholly) [reply util/deltautil/Earthfile:4]: 3-4 poor options :cry: 
* Use relative paths and possibly break other code that depends on this
* Use hashes and break builds locally
* Use version tag and have to update (unexpected results for local runs)
* Copy and paste the 3-4 root Earthfile `+base` commands

I think I prefer the last option.

Comment (alexcb) [reply util/deltautil/Earthfile:4]: I don't see any discussion about why this was done under #2579 and don't see any references to it under our other repos either.

it was initially added as `FROM github.com/earthly/earthly+base`; if instead it was changed to it from a relative import there would be more git history for us to dig into to understand why.

I think we should try using `../..+base` and then if we discover anything breaks then we come back and revert this, and add a comment explaining why.

File: tests/autoskip/arg-change.earth (hunk 2 above)
Comment (alexcb) [reply tests/autoskip/arg-change.earth:3]: > it would be useful to have a way to pin versions

my second idea of using alpine vs busybox would allow this.

```
ARG IMG=alpine:3.19
FROM $IMG
```

then update this:

```
DO --pass-args tests+RUN_EARTHLY --extra_args="--auto-skip-db-path=test.db" --target=+expand-args-from --output_contains="IMG=alpine"
RUN sed -i "s/alpine:.*$/busybox:1.36.1/g" Earthfile
DO --pass-args tests+RUN_EARTHLY --extra_args="--auto-skip-db-path=test.db" --target=+expand-args-from --output_contains="IMG=busybox"
```

Comment (alexcb) [reply tests/autoskip/arg-change.earth:3]: update: my incorrectly understood the `output_contains="VERSION_ID=3.18.6"` part of the test, this requires `cat /etc/os-release`.

Mike and I chatted about this, and instead we can go with the first idea where the version is only specified in `tests/autoskip/Earthfile`, and will use sed to update the value in the `tests/autoskip/arg-change.earth` file

File: tests/autoskip/Earthfile [hunk 4]
```diff
@@ -279,10 +279,15 @@ test-init-failure:
                                      --output_contains="Failed to initialize auto-skip database"
 
 test-arg-change:
+    # Please don't update. These versions are pinned to ensure a version
```
Comment (alexcb) [inline tests/autoskip/Earthfile:282]: I would reword this to be:

```
# this test requires two different versions of alpine
```

then add a following

```
RUN acbtest "$V1" != "$V2"
```

Comment (alexcb) [reply tests/autoskip/Earthfile:282]: the reasoning is to prevent cases where it's somehow auto-updated by an automated process, or user running a `sed` against all Earthfiles, which would miss such a comment.

Comment (mikejholly) [reply tests/autoskip/Earthfile]: @alexcb Updated.

File: util/deltautil/Earthfile (hunk 3 above)
Comment (mikejholly) [reply util/deltautil/Earthfile]: We decided to use relative paths as the full canonical target names did not seem necessary.

**Review by alexcb — APPROVED** [review]
